src/test/scala/lms/thirdparty/test_cuda.scala

```diff
@@ -342,5 +362,96 @@ class CudaTest extends TutorialFunSuite {
   }
+
+  // (Luke) TODO: remove this test case after pytorch one is checked
+  test("kernel_maskedFill_old") {
+    val driver = new DslDriverCCuda[Int, Unit] {
+      @virtualize
+      def snippet(arg: Rep[Int]) = {
+        val n = 4096
+        val input = NewArray[Float](n)
+        val mask = NewArray[Int](n)
+        val output = NewArray[Float](n)
+        val input_d = cudaMalloc2[Float](n)
+        val mask_d = cudaMalloc2[Int](n)
+        val output_d = cudaMalloc2[Float](n)
+
+        for (i <- (0 until n): Rep[Range]) {
+          input(i) = i
+          if (i % 2 == 0) {
+            mask(i) = 1
+          } else {
+            mask(i) = 0
+          }
+        }
+
+        cudaCall(cudaMemcpyOfT[Float](input_d, input, n, host2device))
+        cudaCall(cudaMemcpyOfT[Int](mask_d, mask, n, host2device))
+
+        val maskedFillFloat = maskedFill[Float](true)
+        maskedFillFloat(input_d, output_d, mask_d, 0.0f, (n + 511)/512, 1, 1, 1, n, dim3((n + 511)/512), dim3(512))
+
+        cudaCall(cudaMemcpyOfT[Float](output, output_d, n, device2host))
+
+        printf("masked fill output:\n")
+        for (i <- (0 until n): Rep[Range]) {
+          printf("%f,", output(i))
+        }
+        printf("\n")
+
+        val dinput = NewArray[Float](n)
+        val doutput = NewArray[Float](n)
+        for (i <- (0 until n): Rep[Range]) {
+          doutput(i) = 1.0f
+        }
+        val dinput_d = cudaMalloc2[Float](n)
+        val doutput_d = cudaMalloc2[Float](n)
+
+        cudaCall(cudaMemcpyOfT[Float](doutput_d, doutput, n, host2device))
+
+        val maskedFillGradFloat = maskedFillGrad[Float](true)
+        maskedFillGradFloat(doutput_d, dinput_d, mask_d, (n + 511)/512, 1, 1, 1, n, dim3((n + 511)/512), dim3(512))
+
+        cudaCall(cudaMemcpyOfT[Float](dinput, dinput_d, n, device2host))
+
+        printf("masked fill input gradient:\n")
+        for (i <- (0 until n): Rep[Range]) {
+          printf("%f,", dinput(i))
+        }
+        printf("\n")
+      }
+    }
+  }
+
+  test("kernel_maskedFill") {
+    val driver = new DslDriverCCudeScan[Int, Unit] {
+      
+      @virtualize
+      def snippet(arg: Rep[Int]) = {
+        val d0 = 8
+        val d1 = 8
+        val n = d0 * d1
+        val maskValue = 0.0f
+
+        val input = NewArray[Float](n)
+        scanFile[Float]("golden/maskedFill/input.data", input, n)
+        val cuda_input = cudaMalloc2[Float](n)
+        cudaCall(cudaMemcpyOfT[Float](cuda_input, input, n, host2device))
+
+        val mask = NewArray[Int](n)
+        scanFile[Int]("golden/maskedFill/mask.data", mask, n)
+        val cuda_mask = cudaMalloc2[Int](n)
+        cudaCall(cudaMemcpyOfT[Int](cuda_mask, mask, n, host2device))
+
+        val output = NewArray[Float](n)
+        val cuda_output = cudaMalloc2[Float](n)
+        val maskedFillKernel = maskedFill[Float](true)
+        maskedFillKernel(cuda_input, cuda_output, cuda_mask, maskValue, d0, d1, d0, 1, n, dim3((n + 511)/512), dim3(512))
```

Comment:
I don't think the kernel is called properly :(

Reply:
Yeah, I think the way it's called is slightly inaccurate. Below is a summary of the semantics of maskedFill (we can add this the Scala code as a comment).

```
/*
    Semantics of maskedFill:
    Let's take a n-d tensor input = tensor.randn(10, 2, 5, 8)
    Now let's say we want to apply a mask to the above tensor for dimension 1 and dimension 2;
    Therefore, the shape of the mask should be (2, 5) in this case.
    (if the mask was (5, 2) and needed to apply to dimension 2 and dimension 1, we mark ijSwapped = true (because 2>1))

    We name the two dimensions as dim0 and dim1.
    We need the shape of the two dims (dim0_shape = 2, dim1_shape = 5)
    We also need the strides for the two dims (dim0_stride = 5*8, dim1_stride = 8)
  */
```

Now, in this test case scenario, `ijSwapped` must be `false` because we apply the mask to dimension 0 and 1. And the rest of the params should be as follows;

```
d0_shape = 8
d1_shape = 8
d0_stride = 8 // = d0_shape
d1_stride = 1
```

Can you give this a try and see if it matched the PT output?

Reply:
It worked! Thank you so much for the detailed summary. I understand it now :)